mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -67,12 +67,127 @@ macro_rules! event {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct EndpointConfig {
+    bucket: Option<String>,
+    region: String,
+    use_fips: bool,
+    use_accelerate: bool,
+    use_dual_stack: bool,
+    endpoint: Option<Endpoint>,
+    addressing_style: AddressingStyle,
+}
+
+impl Default for EndpointConfig {
+    fn default() -> Self {
+        Self {
+            bucket: None,
+            region: "us-east-1".to_owned(),
+            use_fips: false,
+            use_accelerate: false,
+            use_dual_stack: false,
+            endpoint: None,
+            addressing_style: AddressingStyle::Automatic,
+        }
+    }
+}
+
+impl EndpointConfig {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set the bucket name/alias/arn for endpoint configuration
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn bucket(mut self, bucket: &str) -> Self {
+        self.bucket = Some(bucket.to_owned());
+        self
+    }
+
+    // Set the region for endpoint configuration
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn region(mut self, region: &str) -> Self {
+        self.region = region.to_owned();
+        self
+    }
+
+    /// use FIPS config for S3
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn use_fips(mut self, fips: bool) -> Self {
+        self.use_fips = fips;
+        self
+    }
+
+    /// use Transfer Acceleration config for S3
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn use_accelerate(mut self, accelerate: bool) -> Self {
+        self.use_accelerate = accelerate;
+        self
+    }
+
+    /// use dual stack config for S3
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn use_dual_stack(mut self, dual_stack: bool) -> Self {
+        self.use_dual_stack = dual_stack;
+        self
+    }
+
+    /// Set predefined url for endpoint configuration
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn endpoint(mut self, endpoint: Endpoint) -> Self {
+        self.endpoint = Some(endpoint);
+        self
+    }
+
+    /// Set addressing style for endpoint configuration
+    #[must_use = "EndpointConfig follows a builder pattern"]
+    pub fn addressing_style(mut self, addressing_style: AddressingStyle) -> Self {
+        self.addressing_style = addressing_style;
+        self
+    }
+
+    /// Get bucket name if set for endpoint resolution
+    pub fn get_bucket(&self) -> Option<String> {
```

Comment:
Do any of these getters need to be called anywhere other than the endpoint resolver? I think you end up just adding a `resolve_for_bucket` method to this `EndpointConfig`, and don't need these getters any more.

Reply:
Sure. But, the endpoint can be configured even without a bucket. And since this is mountpoint-s3-client, shouldn't we keep it more generic ratehr than resolving endpoint only when bucket is provided. I agree that we should add the bucket information for each S3 request.

Reply:
Yes. But I am keeping the setters like `pub fn endpoint(mut self, endpoint: Endpoint) -> Self` to set the config in the `main.rs` .